Fix program closing when map changes

AnomalousNicole opened the pull request:
I would like to start by thanking James Turner for creating Borderless1942 for my favorite game Battlefield 1942.

This is my first GitHub pull request, so please let me know if I did something incorrectly. :)

Recently while playing Battlefield 1942 using Borderless1942, I discovered the program would close during a multiplayer map change.  Since I enjoy programming, I decided to dig into the code and figure out what was happening.  Below is the small modification that I made to prevent the Borderless1942 program from closing when changing maps during multiplayer.

I tested the changes locally in COOP mode using the master.bf1942.org master server patch from the link below.
https://www.moddb.com/games/battlefield-1942/downloads/bf1942-masterbf1942org-patch

AnomalousNicole committed:
Small modification to prevent the Borderless1942 program from closing when changing maps during multiplayer.

Tested the changes using the master.bf1942.org master server patch from the link below.
https://www.moddb.com/games/battlefield-1942/downloads/bf1942-masterbf1942org-patch

Turnerj commented on Program.cs:
```diff
@@ -4,15 +4,18 @@
 var processStartInfo = new ProcessStartInfo
 {
-	FileName = @"BF1942.exe"
+    FileName = @"BF1942.exe"
```
I know this may seem pretty minor but try to avoid whitespace changes (in this case, changing from tabs to spaces). There are a few cases where this has happened in this file that would be good to undo.

AnomalousNicole replied:
Even though it is minor, thank you for bringing this change to my attention.  I think have probably developed a bad habit of using the CTRL+K, D functionality in VS 2022 with the default settings.  It sounds like this is a habit that I should break. :)  In my previous job when I was coding in the Borland Delphi 5 IDE, there wasn't any functionality similar to the CTRL+K, D command.  I will just have to train myself not to use the CTRL+K, D functionality in VS 2022 or maybe there is another setting/option that I can use.

Turnerj replied:
To be fair, `CTRL+K, D` is generally a pretty good habit... if I had an `.editorconfig` file that defined the result I wanted (VS2022 will follow whatever is defined there).

Some of my projects have it, some don't - probably should fix that up.

Turnerj commented on Program.cs:
```diff
@@ -4,15 +4,18 @@
+
 var keepAlive = true;
-Console.WriteLine($"Process has started ({process.Id})");
+Console.WriteLine($"[Borderless1942]: [{DateTime.Now:MM/dd/yyyy - hh:mm:ss tt}] [BF1942 Process Has Started]: ({process.Id})");
```
I like the more detailed log messages and am generally happy with them. One thing I would switch up is the date format.

Because different countries have different date format rules, I find it best to use `yyyy-MM-dd` as it removes ambiguity between this program running on your computer and it running on mine. (In case you're curious, `yyyy-MM-dd` is part of a standard called ISO 8601)

AnomalousNicole replied:
That is definitely a good point regarding the date formatting.  Thank you for pointing out the ISO 8601 standard.  I will keep that in mind for any future contributions or projects that I write code for.  Unfortunately, I seem to be stuck in a mindset of using the MM/dd/yyyy format since that is the format that I have used for my of my career so far.  I need to train myself to think more about the end user experience and that they may not be using the same regional settings that I am.

Turnerj commented on Program.cs:
```diff
@@ -22,37 +25,47 @@
+            {
+                process = matchingProcesses[0];
+
+                Console.WriteLine($"[Borderless1942]: [{DateTime.Now:MM/dd/yyyy - hh:mm:ss tt}] [BF1942 Process Has Changed] [Map Change Detected] ({oldProcessId} -> {process.Id})");
```
While this definitely does happen on map change, it isn't the only reason it happens. Just navigating from in-game to the main menu switches process (which is just weird in my opinion but here we are). It is partly why I was ambiguous about the cause of the process change.

So besides my comment about date format, I'd probably just say "Process Change Detected" instead of specifically "Map Change Detected".

AnomalousNicole replied:
I was not aware that the process also changed when navigating from in-game to the main menu.  Although now that you mention that, I do seem to remember that you mentioned that fact in your blog post.  I will remove the map change detected text from the console message.

Turnerj commented on Program.cs:
```diff
@@ -22,37 +25,47 @@
+
+                Console.WriteLine($"[Borderless1942]: [{DateTime.Now:MM/dd/yyyy - hh:mm:ss tt}] [BF1942 Process Has Changed] [Map Change Detected] ({oldProcessId} -> {process.Id})");
+
+                if (oldProcessId == process.Id)
```
I don't think this would be possible to hit. At this point, the previous process we were tracking has closed and we just found a new BF1942 process. It is very unlikely to have the same process Id given out by the OS.

What I think you're trying to get at with the core concept of your changes is primarily increasing how many times we retry finding the process.

The existing code worked by looking for the new process for 2 seconds (2 retries x 1 second wait) which sounds like it might not be long enough. I'd suggest probably then increasing `retryCount < 2` to be `retryCount < 3` effectively giving another second of time before Borderless1942 closes itself.

Part of the reason for the auto-close mechanism of Borderless1942 was convenience but it was trying to weigh that between auto-closing and still detecting change in process of BF1942. Really though, if it is causing more problems with auto-close, we could alternatively remove the auto-close logic entirely though we'd want to include a log message for every time we can't find the BF1942 process. Something like "Attempting to find next BF1942 process" etc.

AnomalousNicole replied:
Thank you for explaining your thought process behind the auto-close functionality and the BF1942 process detection.
I made the changes below to the program and did some more testing locally of the changes.
Below is an explanation of the changes and the steps I used to test them.
Sorry for the wall of text below, but I wanted to be detailed in describing what I changed and how I tested the changes. :)


Code changes:
I changed "while (retryCount < 2)" to "while (retryCount < 3)".

I updated the console messages to use the ISO 8601 date format.

I left in the "if (oldProcessId == process.Id)" condition but modified the body to call "Environment.Exit(0)" instead of breaking out of the "while (retryCount < 3)" loop.
The testing from EXE 2 provides an explanation for why I left in the "if (oldProcessId == process.Id)" condition.
I think when I put the break in initially, I was meaning for the code to break out of the MainLoop loop and now realize that the code was breaking out of the retryCount loop instead.

It may not be necessary to set keepAlive to false before calling Environment.Exit(0).
I would have to give it some more thought, to come up with something that could be used to keep the program running until the "if (oldProcessId == process.Id)" is hit.


I created two Borderless1942 EXE files, one with the "if (oldProcessId == process.Id)" change present and one with the change commented out.
Using EXE 1 with "if (oldProcessId == process.Id)" present, Borderless1942 automatically closed when BF1942 was closed from the main menu.
Using EXE 2 with "if (oldProcessId == process.Id)" commented out, Borderless1942 did not close automatically when BF1942 was closed from the main menu.

Below is the output from each EXE.
EXE 1: "if (oldProcessId == process.Id)" Uncommented
[Borderless1942]: [2022-08-20 - 10:02:47 AM] [BF1942 Process Has Started] [36724]
 When Battle of Britain loaded, BF1942 did not create a new process for the first map in the playlist.

[Borderless1942]: [2022-08-20 - 10:04:12 AM] [BF1942 Process Has Changed] [36724 -> 41068]
 When Battle of Midway loaded, BF1942 created the process above.	

[Borderless1942]: [2022-08-20 - 10:05:15 AM] [BF1942 Process Has Changed] [41068 -> 40264]
 When Battle of the Bulge loaded, BF1942 created the process above.

[Borderless1942]: [2022-08-20 - 10:05:44 AM] [BF1942 Process Has Exited] [Closing Program] [40264 -> 40264]
 When I hit ESC and closed the game from the main menu, Borderless1942 closed and output the console message above.


EXE 2: "if (oldProcessId == process.Id)" Commented out
[Borderless1942]: [2022-08-20 - 08:49:35 AM] [BF1942 Process Has Started] [38916]
 When Battle of Britain loaded, BF1942 did not create a new process for the first map in the playlist.

[Borderless1942]: [2022-08-20 - 08:50:41 AM] [BF1942 Process Has Changed] [38916 -> 39376]
 When Battle of Midway loaded, BF1942 created the process above.

[Borderless1942]: [2022-08-20 - 08:51:15 AM] [BF1942 Process Has Changed] [39376 -> 18892]
 When Battle of the Bulge loaded, BF1942 created the process above.

[Borderless1942]: [2022-08-20 - 08:51:39 AM] [BF1942 Process Has Changed] [18892 -> 18892]
 When I hit ESC and closed the game from the main menu, Borderless1942 output the console message above.
 It seems that the code goes into an infinite loop condition, when BF1942 is closed from the main menu.
 When "await process.WaitForMainWindowAsync()" is called from the body of "if (matchingProcesses.Length == 1)", Borderless1942 loops infinitely on the "while (process.MainWindowHandle == IntPtr.Zero)" condition since the process.MainWindowHandle is equal to IntPtr.Zero.


Testing Steps:
Under Windows 11, I ran the following tests in a local COOP game using the following steps.

1. I opened a Terminal window in the root of my BF1942 install and ran ".\Borderless1942.exe" so that I would be able to see all of the console messages.

2. I created a local COOP mode playlist with the maps below.
Battle of Britain
Battle of Midway
Battle of the Bulge

3. To switch to the next map in the playlist, I used the in game map vote command by holding Insert and selecting the next map in the list.

4. As expected, BF1942 created a new process every time the map was changed except for when Battle of Britain loaded.

5. After Battle of the Bulge loaded, I hit ESC and then closed the game without navigating to any other parts of the menu.

Turnerj replied:
Thanks for the details!

So I think the strange thing for me is that `Process.GetProcessesByName("BF1942")` is returning the _old_ process after it has exited.

> It seems that the code goes into an infinite loop condition, when BF1942 is closed from the main menu. When "await process.WaitForMainWindowAsync()" is called from the body of "if (matchingProcesses.Length == 1)", Borderless1942 loops infinitely on the "while (process.MainWindowHandle == IntPtr.Zero)" condition since the process.MainWindowHandle is equal to IntPtr.Zero.

Yeah, I could definitely see that happening if we're calling it on an already closed process. It is strange though _this_ problem didn't exist in my original code though.

One thought I have is perhaps trying again with `if (oldProcessId == process.Id)` commented out but also re-instating the change you removed here: https://github.com/Turnerj/Borderless1942/pull/1/files#diff-0b69b473fe937040615d69f606751f61ddbc2e3a1849360ff2456c22afe88c0bL46

Remember that the code is meant to retry a certain number of times. Without the `keepAlive = false;` line after the retry loop, we are basically giving the whole thing another chance to retry which doesn't make a lot of sense.

AnomalousNicole replied:
@Turnerj You're welcome. :)
Below I've outlined the latest changes I've made and the steps used in testing them.


**Code changes:**
I removed the "if (oldProcessId == process.Id)" block of code.
I added a new check to see if BF1942 has been closed from the main menu and if the MainWindowHandle is equal to IntPtr.Zero or not.  I then called break to break out of the retryCount while loop, so that keepAlive is set to false and the program closes.  When BF1942 is closed from the main menu, I don't believe the code should call "process.WaitForMainWindowAsync" since it causes an infinite loop condition.

Do you recall which BF1942 exe you used during your initial development of Borderless1942?
I guess it's possible that some of the latest BF1942 exe modifications might be playing a role in the behavior we are seeing now.

Here are the notable changes from the latest BF1942 exe:  The latest exe can be found below.
https://www.moddb.com/games/battlefield-1942/downloads/bf1942-masterbf1942org-patch

* No CD
* Portable
* Widescreen support - In game
* Widescreen support - In menu
* master.gamespy.com replaced with master.bf1942.org
* Increased accessible ram from 2gb to 4gb
* Removed this message from console:
 Error: failed  to add player <name> to buddylist. Couldn't locate player.


**Debugging output:**
Below is the debug console output from the latest round of code changes.
I removed the date time output from the debug console output, but left it in the committed code change.

[Borderless1942]: [BF1942 Process Has Started] [55420]
When Battle of Britain loaded, BF1942 did not create a new process for the first map in the playlist.

[Borderless1942]: [BF1942 Process Has Changed] [55420 -> 56984] -> [process.MainWindowHandle = 85727216] -> [process.HasExited = False]
When Battle of Midway loaded, BF1942 created the process above.  The MainWindowHandle is not equal to IntPtr.Zero and HasExited is False.

[Borderless1942]: [BF1942 Process Has Changed] [56984 -> 41716] -> [process.MainWindowHandle = 32642698] -> [process.HasExited = False]
When Battle of the Bulge loaded, BF1942 created the process above.  The MainWindowHandle is not equal to IntPtr.Zero and HasExited is False.

[Borderless1942]: [BF1942 Process Has Changed] [41716 -> 41716] -> [process.MainWindowHandle = 0] -> [process.HasExited = True]
[Borderless1942]: [BF1942 Process Has Exited]
When I hit ESC and closed the game from the main menu, Borderless1942 closed and output the console messages above.


**Testing Steps:**
Under Windows 11, I ran the following tests in a local COOP game using the following steps.
1. I opened a Terminal window in the root of my BF1942 install and ran ".\Borderless1942.exe" so that I would be able to see all of the console messages.

2. I created a local COOP mode playlist with the maps below.
Battle of Britain
Battle of Midway
Battle of the Bulge

3. To switch to the next map in the playlist, I used the in game map vote command by holding Insert and selecting the next map in the list.

4. As expected, BF1942 created a new process every time the map was changed except for when Battle of Britain loaded.

5. After Battle of the Bulge loaded, I hit ESC and then closed the game without navigating to any other parts of the menu.

Turnerj replied:
Hey @johnbrechbuhl85 - I'm using the 1.61 patched version by [Team Simple](https://team-simple.org/download/).

After some messing around with `Process.GetProcessesByName()`, I've hit the same issue you're trying to do but with other programs. That API is basically no reliable for getting _running_ processes without also filtering by `HasExited`. I've commented on your change with the tweaks I'd like before we can merge it in.

Have a mess around with it and see if it works well for you 🙂

AnomalousNicole replied:
The Time Simple version should be fairly similar to the latest patched exe I linked to.  The major change was the master server address patch in the latest exe.

Turnerj reviewed (changes requested):
Hey @johnbrechbuhl85 - thanks for raising this pull request! It is awesome that your first open source contribution is to this little ol' project. 🙂

I've added some comments about a few things. The main problem really is how it affects the auto-closing behaviour of Borderless1942.

Happy to walk you through the details of my comments if you have any questions.

AnomalousNicole committed:
Reverted white space changes and updated console messages to use ISO 8601 date format.

Modified the "if (oldProcessId == process.Id)" condition to call Environment.Exit(0) and close Borderless1942 when BF1942 is closed.

AnomalousNicole committed:
Check to see if BF1942 has been closed from the main menu and if the MainWindowHandle is equal to IntPtr.Zero or not.

If those conditions are met, then break out of the retryCount loop and set keepAlive to false.

Turnerj commented on Program.cs:
```diff
@@ -29,13 +29,17 @@
 				process = matchingProcesses[0];
-				Console.WriteLine($"Process has changed ({oldProcessId} -> {process.Id})");
+				Console.WriteLine($"[Borderless1942]: [{DateTime.Now:yyyy-MM-dd - hh:mm:ss tt}] [BF1942 Process Has Changed] [{oldProcessId} -> {process.Id}]");
+				if (process.HasExited && process.MainWindowHandle == IntPtr.Zero)
+				{
+					break;
+				}
```
OK, after debugging `Process.GetProcessesByName()`, I've been able to separately confirm that it can return closed processes. Here's what I think we should do:
- Drop the `process.MainWindowHandle` check
- Move the if-statement before the log output

AnomalousNicole replied:
Hey @Turnerj,

Below I've outlined the latest changes I've made and the steps used in testing them.
Borderless1942 appears to be functioning correctly with the latest round of changes.

**Code changes:**
I made the requested changes to the body of "if (matchingProcesses.Length == 1)".


**Debugging output:**
Below is the debug console output from the latest round of code changes.

[Borderless1942]: [2022-08-27 - 08:39:34 AM] [BF1942 Process Has Started] [47108]
When Battle of Britain loaded, BF1942 did not create a new process for the first map in the playlist.

[Borderless1942]: [2022-08-27 - 08:40:57 AM] [BF1942 Process Has Changed] [47108 -> 46212]
When Battle of Midway loaded, BF1942 created the process above.

[Borderless1942]: [2022-08-27 - 08:42:01 AM] [BF1942 Process Has Changed] [46212 -> 36248]
When Battle of the Bulge loaded, BF1942 created the process above.

[Borderless1942]: [2022-08-27 - 08:42:31 AM] [BF1942 Process Has Exited]
When I hit ESC and closed the game from the main menu, Borderless1942 closed and output the console messages above.


**Testing Steps:**
Under Windows 11, I ran the following tests in a local COOP game using the following steps.

1. I opened a Terminal window in the root of my BF1942 install and ran ".\Borderless1942.exe" so that I would be able to see all of the console messages.

2. I created a local COOP mode playlist with the maps below.
Battle of Britain
Battle of Midway
Battle of the Bulge

3. To switch to the next map in the playlist, I used the in game map vote command by holding Insert and selecting the next map in the list.

4. As expected, BF1942 created a new process every time the map was changed except for when Battle of Britain loaded.

5. After Battle of the Bulge loaded, I hit ESC and then closed the game without navigating to any other parts of the menu.

AnomalousNicole committed:
Moved the process.HasExited check before the Process Has Changed message and removed the process.MainWindowHandle == IntPtr.Zero check.

Turnerj commented:
Looks good mate - thanks for your work on this!

AnomalousNicole commented:
You're welcome. :)  It was a good learning experience and I enjoyed working on the changes.  I'm glad that I was able to contribute to the app.
